controlplane/kubeadm/internal/workload_cluster_coredns_test.go

```diff
@@ -287,62 +291,75 @@ kind: ClusterConfiguration
-			Client:          testEnv.GetClient(),
-			CoreDNSMigrator: tt.migrator,
-		}
-		err := w.UpdateCoreDNS(ctx, tt.kcp)
-		if tt.expectErr {
-			g.Expect(err).To(HaveOccurred())
-			return
-		}
-		g.Expect(err).ToNot(HaveOccurred())
+		t.Run(tt.name, func(t *testing.T) {
+			for _, o := range tt.objs {
+				// NB. deep copy test object so changes applied during a test does not affect other tests.
+				o := o.DeepCopyObject().(client.Object)
+				g.Expect(testEnv.Create(ctx, o)).To(Succeed())
+				// this makes sure that the cache is updated with the object
+				// to avoid 404 errors leading to test flakes
+				g.Eventually(func() bool {
+					key, _ := client.ObjectKeyFromObject(o)
+					err := testEnv.Get(ctx, key, o)
+					if err != nil && apierrors.IsNotFound(err) {
+						return false
+					}
+					return true
+				}, "10s").Should(BeTrue())
+			}
 
-		// Assert that CoreDNS updates have been made
-		if tt.expectUpdates {
-			// assert kubeadmConfigMap
-			var expectedKubeadmConfigMap corev1.ConfigMap
-			g.Expect(testEnv.Get(ctx, ctrlclient.ObjectKey{Name: kubeadmConfigKey, Namespace: metav1.NamespaceSystem}, &expectedKubeadmConfigMap)).To(Succeed())
-			g.Expect(expectedKubeadmConfigMap.Data).To(HaveKeyWithValue("ClusterConfiguration", ContainSubstring("1.7.2")))
-			g.Expect(expectedKubeadmConfigMap.Data).To(HaveKeyWithValue("ClusterConfiguration", ContainSubstring("k8s.gcr.io/some-repo")))
+			// Register cleanup function
+			t.Cleanup(func() {
+				// Cleanup test objects (and wait for deletion to complete).
+				_ = testEnv.Cleanup(ctx, tt.objs...)
+				g.Eventually(func() bool {
+					for _, o := range []client.Object{cm, depl, kubeadmCM, badCM} {
+						key, _ := client.ObjectKeyFromObject(o)
+						err := testEnv.Get(ctx, key, o)
+						if err == nil || (err != nil && !apierrors.IsNotFound(err)) {
+							return false
+						}
+					}
+					return true
+				}, "10s").Should(BeTrue())
+			})
 
-			// assert CoreDNS corefile
-			var expectedConfigMap corev1.ConfigMap
-			g.Expect(testEnv.Get(ctx, ctrlclient.ObjectKey{Name: coreDNSKey, Namespace: metav1.NamespaceSystem}, &expectedConfigMap)).To(Succeed())
-			g.Expect(expectedConfigMap.Data).To(HaveLen(2))
-			g.Expect(expectedConfigMap.Data).To(HaveKeyWithValue("Corefile", "updated-core-file"))
-			g.Expect(expectedConfigMap.Data).To(HaveKeyWithValue("Corefile-backup", expectedCorefile))
+			w := &Workload{
+				Client:          testEnv.GetClient(),
+				CoreDNSMigrator: tt.migrator,
+			}
+			err := w.UpdateCoreDNS(ctx, tt.kcp)
 
-			// assert CoreDNS deployment
-			var actualDeployment appsv1.Deployment
-			g.Eventually(func() string {
-				g.Expect(testEnv.Get(ctx, ctrlclient.ObjectKey{Name: coreDNSKey, Namespace: metav1.NamespaceSystem}, &actualDeployment)).To(Succeed())
-				return actualDeployment.Spec.Template.Spec.Containers[0].Image
-			}, "5s").Should(Equal("k8s.gcr.io/some-repo/coredns:1.7.2"))
-		}
+			if tt.expectErr {
+				g.Expect(err).To(HaveOccurred())
+			} else {
```

Comment:
I think you may need to undo this - if we expect an error, we don't want to make any more assertions (like tt.expectedUpdates), do we?

Reply:
Yes you are right, we should not need to run any more assertions in case of errors. And in all the test inputs listed with an error expected, the `expectedUpdates` is unset.
But if `return` makes it more clear I will add it back. I initially removed it to make sure clean up gets registered, and then thought against adding it back.